typescript/examples/flatbufferswriter/scripts/flatbufferUtils.ts

```diff
@@ -0,0 +1,118 @@
+
+import {
+  FrameTransform as FbFrameTransform,
+  Quaternion as FbQuaternion,
+  Time as FbTime,
+  Vector3 as FbVector3,
+} from "../../flatbuffer/output/FrameTransform_generated";
+import {
+  PackedElementField as FbPackedElementField,
+  Pose as FbPose,
+  Grid as FbGrid,
+  Vector2 as FbVector2,
+} from "../../flatbuffer/output/Grid_generated";
+
+export function buildQuaternion(builder: Builder, quatJson: Quaternion): number {
+  FbQuaternion.startQuaternion(builder);
+  FbQuaternion.addX(builder, quatJson.x);
+  FbQuaternion.addY(builder, quatJson.y);
+  FbQuaternion.addZ(builder, quatJson.z);
+  FbQuaternion.addW(builder, quatJson.w);
+  return FbQuaternion.endQuaternion(builder);
+}
+
+export function buildVector2(builder: Builder, json: Vector2): number {
+  FbVector2.startVector2(builder);
+  FbVector2.addX(builder, json.x);
+  FbVector2.addY(builder, json.y);
+  return FbVector2.endVector2(builder);
+}
+
+export function buildVector3(builder: Builder, json: Vector3): number {
+  FbVector3.startVector3(builder);
+  FbVector3.addX(builder, json.x);
+  FbVector3.addY(builder, json.y);
+  FbVector3.addZ(builder, json.z);
+  return FbVector3.endVector3(builder);
+}
+
+export function buildPose(builder: Builder, json: Pose): number {
+  const pos = buildVector3(builder, json.position);
+  const quat = buildQuaternion(builder, json.orientation);
+  FbPose.startPose(builder);
+  FbPose.addOrientation(builder, quat);
+  FbPose.addPosition(builder, pos);
+  return FbPose.endPose(builder);
+}
+
+export function buildTfMessage(builder: Builder, tfJson: FrameTransform): number {
+  const parentFrameId = builder.createString(tfJson.parent_frame_id);
+  const childFrameId = builder.createString(tfJson.child_frame_id);
+
+  const quat = buildQuaternion(builder, tfJson.rotation);
+
+  const vec3 = buildVector3(builder, tfJson.translation);
+
+  FbFrameTransform.startFrameTransform(builder);
+  FbFrameTransform.addParentFrameId(builder, parentFrameId);
+  FbFrameTransform.addChildFrameId(builder, childFrameId);
+  FbFrameTransform.addTranslation(builder, vec3);
+  FbFrameTransform.addRotation(builder, quat);
+  FbFrameTransform.addTimestamp(
+    builder,
+    FbTime.createTime(builder, BigInt(tfJson.timestamp.sec), tfJson.timestamp.nsec),
+  );
+
+  const tf = FbFrameTransform.endFrameTransform(builder);
+  return tf;
+}
+
+export function buildPackedElementField(builder: Builder, json: PackedElementField): number {
+  const name = builder.createString(json.name);
+  FbPackedElementField.startPackedElementField(builder);
+  FbPackedElementField.addName(builder, name);
+  FbPackedElementField.addOffset(builder, json.offset);
+  FbPackedElementField.addType(builder, json.type);
+  return FbPackedElementField.endPackedElementField(builder);
+}
+
+export function buildGridMessage(builder: Builder, json: Grid): number {
+  const frameId = builder.createString(json.frame_id);
+
+  const pose = buildPose(builder, json.pose);
```

Comment:
The builder as far as I cant tell allows for abstracted use of the underlying buffer, expansion of buffer and tracks how much of it is allocated. So when you start a new object on the buffer and add attributes to it it allocates a given space for it in the continuous chunk and tracks where the next object can begin. It also allows the buffer to have a `root` schema, hence the importance of ending the buffer with the main message type. All of the fields and data in a message have to be in the same buffer so when you create/end a `Quaternion` the index of that quaternion needs to be used when adding it to the object of the schema that includes it so that it can reference it in the same buffer.